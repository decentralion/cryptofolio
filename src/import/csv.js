// @flow

import moment from "moment";
import Big from "big.js";
import parse from "csv-parse/lib/sync";
import {parseNumber, parseDate} from "./parse";

import type {Transaction} from "../core/transaction";
import {validate} from "../core/transaction";

const TX_SOURCE = "CAT_CSV_IMPORT";

export function parseTransactions(csvData: string) {
  const csvRows = parse(csvData);
  csvRows.shift();
  return csvRows.map((r) => {
    const tx = {
      date: parseDate(r[0]),
      ticker: r[1],
      amount: parseNumber(r[2]),
      type: r[3],
      price: parseNumber(r[4]),
      txSource: TX_SOURCE,
    };
    validate(tx);
    return tx;
  });
}

export function parseTrades(csvData: string) {
  const csvRows = parse(csvData);
  csvRows.shift();
  const transactions = [];
  csvRows.forEach((r) => {
    const date = parseDate(r[0]);
    const srcTicker = r[1];
    const srcAmount = parseNumber(r[2]);
    const srcPrice = parseNumber(r[3]);
    const dstTicker = r[4];
    const dstAmount = parseNumber(r[5]);
    const dstPrice = parseNumber(r[6]);
    const sell: Transaction = {
      ticker: srcTicker,
      price: srcPrice,
      amount: srcAmount.times(-1),
      date,
      type: "TRADE",
      txSource: TX_SOURCE,
    };
    validate(sell);
    transactions.push(sell);
    const buy: Transaction = {
      ticker: dstTicker,
      price: dstPrice,
      amount: dstAmount,
      date,
      type: "TRADE",
      txSource: TX_SOURCE,
    };
    validate(buy);
    transactions.push(buy);
  });
  return transactions;
}
