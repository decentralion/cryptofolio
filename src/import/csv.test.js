// @flow

import {parseTrades, parseTransactions} from "./csv";

describe("parseTrades", () => {
  const tradesTestCase = `\
Date,Src,Src Amount,Src Price,Dst,Dst Amount,Dst Price
9/29/2010,USD,"4,000.000",$1.00,BTC,1.000,"$4,000.00"
10/1/2010,USD,"2,000.000",$1.00,ETH,6.000,$300.00
10/1/2010,USD,600.000,$1.00,ETH,2.000,$300.00
10/1/2010,USD,"1,000.000",$1.00,LTC,20.000,$50.00`;
  it("snapshots as expected", () => {
    expect(parseTrades(tradesTestCase)).toMatchSnapshot();
  });
});

describe("parseTransactions", () => {
  const transactionsTestCase = `\
Date,Currency,Amount,Type,Price On Date,Notes
1/1/2010,BTC,2.0101,FUND,$250.00,Start
8/1/2017,BCH,2.0101,FORK,$500.00,BCH Fork`;
  it("snapshots as expected", () => {
    expect(parseTransactions(transactionsTestCase)).toMatchSnapshot();
  });
});
