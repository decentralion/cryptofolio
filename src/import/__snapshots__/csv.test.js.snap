// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseTrades snapshots as expected 1`] = `
Array [
  Object {
    "amount": "-4000",
    "date": "2010-09-29T07:00:00.000Z",
    "price": "1",
    "ticker": "USD",
    "txSource": "CAT_CSV_IMPORT",
    "type": "TRADE",
  },
  Object {
    "amount": "1",
    "date": "2010-09-29T07:00:00.000Z",
    "price": "4000",
    "ticker": "BTC",
    "txSource": "CAT_CSV_IMPORT",
    "type": "TRADE",
  },
  Object {
    "amount": "-2000",
    "date": "2010-10-01T07:00:00.000Z",
    "price": "1",
    "ticker": "USD",
    "txSource": "CAT_CSV_IMPORT",
    "type": "TRADE",
  },
  Object {
    "amount": "6",
    "date": "2010-10-01T07:00:00.000Z",
    "price": "300",
    "ticker": "ETH",
    "txSource": "CAT_CSV_IMPORT",
    "type": "TRADE",
  },
  Object {
    "amount": "-600",
    "date": "2010-10-01T07:00:00.000Z",
    "price": "1",
    "ticker": "USD",
    "txSource": "CAT_CSV_IMPORT",
    "type": "TRADE",
  },
  Object {
    "amount": "2",
    "date": "2010-10-01T07:00:00.000Z",
    "price": "300",
    "ticker": "ETH",
    "txSource": "CAT_CSV_IMPORT",
    "type": "TRADE",
  },
  Object {
    "amount": "-1000",
    "date": "2010-10-01T07:00:00.000Z",
    "price": "1",
    "ticker": "USD",
    "txSource": "CAT_CSV_IMPORT",
    "type": "TRADE",
  },
  Object {
    "amount": "20",
    "date": "2010-10-01T07:00:00.000Z",
    "price": "50",
    "ticker": "LTC",
    "txSource": "CAT_CSV_IMPORT",
    "type": "TRADE",
  },
]
`;

exports[`parseTransactions snapshots as expected 1`] = `
Array [
  Object {
    "amount": "2.0101",
    "date": "2010-01-01T08:00:00.000Z",
    "price": "250",
    "ticker": "BTC",
    "txSource": "CAT_CSV_IMPORT",
    "type": "FUND",
  },
  Object {
    "amount": "2.0101",
    "date": "2017-08-01T07:00:00.000Z",
    "price": "500",
    "ticker": "BCH",
    "txSource": "CAT_CSV_IMPORT",
    "type": "FORK",
  },
]
`;
